parquet-variant/src/main/java/org/apache/parquet/variant/Variant.java

```diff
@@ -0,0 +1,313 @@
+  }
+
+  /**
+   * @return the string value
+   */
+  public String getString() {
+    return VariantUtil.getString(value);
+  }
+
+  /**
+   * The value type of Variant value. It is determined by the header byte.
+   */
+  public enum Type {
+    OBJECT,
+    ARRAY,
+    NULL,
+    BOOLEAN,
+    BYTE,
+    SHORT,
+    INT,
+    LONG,
+    STRING,
+    DOUBLE,
+    DECIMAL4,
+    DECIMAL8,
+    DECIMAL16,
+    DATE,
+    TIMESTAMP_TZ,
+    TIMESTAMP_NTZ,
+    FLOAT,
+    BINARY,
+    TIME,
+    TIMESTAMP_NANOS,
+    TIMESTAMP_NANOS_NTZ,
+    UUID
+  }
+
+  /**
+   * @return the type of the variant value
+   */
+  public Type getType() {
+    return VariantUtil.getType(value);
+  }
+
+  /**
+   * @return the number of object fields in the variant. `getType()` must be `Type.OBJECT`.
+   */
+  public int numObjectElements() {
+    return VariantUtil.getObjectInfo(value).numElements;
+  }
+
+  /**
+   * Returns the object field Variant value whose key is equal to `key`.
+   * Return null if the key is not found. `getType()` must be `Type.OBJECT`.
+   * @param key the key to look up
+   * @return the field value whose key is equal to `key`, or null if key is not found
+   */
+  public Variant getFieldByKey(String key) {
+    VariantUtil.ObjectInfo info = VariantUtil.getObjectInfo(value);
+    // Use linear search for a short list. Switch to binary search when the length reaches
+    // `BINARY_SEARCH_THRESHOLD`.
+    if (info.numElements < BINARY_SEARCH_THRESHOLD) {
+      for (int i = 0; i < info.numElements; ++i) {
+        ObjectField field = getFieldAtIndex(
+            i,
+            value,
+            metadata,
+            info.idSize,
+            info.offsetSize,
+            value.position() + info.idStartOffset,
+            value.position() + info.offsetStartOffset,
+            value.position() + info.dataStartOffset);
+        if (field.key.equals(key)) {
+          return field.value;
+        }
+      }
+    } else {
+      int low = 0;
+      int high = info.numElements - 1;
+      while (low <= high) {
+        // Use unsigned right shift to compute the middle of `low` and `high`. This is not only a
+        // performance optimization, because it can properly handle the case where `low + high`
+        // overflows int.
+        int mid = (low + high) >>> 1;
+        ObjectField field = getFieldAtIndex(
+            mid,
+            value,
+            metadata,
+            info.idSize,
+            info.offsetSize,
+            value.position() + info.idStartOffset,
+            value.position() + info.offsetStartOffset,
+            value.position() + info.dataStartOffset);
+        int cmp = field.key.compareTo(key);
+        if (cmp < 0) {
+          low = mid + 1;
+        } else if (cmp > 0) {
+          high = mid - 1;
+        } else {
+          return field.value;
+        }
+      }
+    }
+    return null;
+  }
+
+  /**
+   * A field in a Variant object.
+   */
+  public static final class ObjectField {
```

Comment:
I don't think this needs to be public because it isn't returned through the public methods.